docs/community/20241112-node-bootstrapping.md

```diff
@@ -0,0 +1,117 @@
+cleanly integrate my contributions into the node bootstrapping process without
+interfering with other implementations.
+I would like my contributions to be as re-usable as possible across different
+bootstrap providers.
+I would like to read and to follow documentation, guidelines, and specifications
+on the above.
+I would like to offer a choice of node bootstrapping configuration systems to users,
+enable distribution vendors to extend the choces available by adding new
+configuration systems, and ease maintenance and enable extension of currently
+supported systems for the ClusterAPI developer community.
+
+
+**Problem statement / Example issues**
+
+Currently, the kubeadm bootstrap provider implements two provisioning systems:
+cloud-init and ignition.
+As there is currently no abstract way to pass configuration to  node bootstrapping,
+both configuration system implementations - ignition and cloud-init - must cover
+generic configuration as well as config language specific concerns.
+This leads to duplicate code and to maintenance overhead.
+Also, other providers like e.g. MicroK8s must implement their own OS configuration
+generators since kubeadm's node bootstrapping implementations cannot be re-used.
+MicroK8s currently only supports cloud-init.
+
+Outside of ClusterAPI, both [cloud-init](https://github.com/canonical/cloud-init)
+and [ignition](https://github.com/coreos/ignition) provisioning are widely adopted
+across distributions.
+While cloud-init is used by general-purpose Linux distributions like Ubuntu/Debian,
+SUSE Linux, Alma, Rocky, Fedora, and Red Hat Enterprise Linux, ignition is popular
+with special-purpose distributions like Fedora CoreOS / Red Hat CoreOS, SuSE MicroOS
+/ SLE Micro / Kalpa, and Flatcar Container Linux.
+It is likely that more provisioning systems exist; a clean separation between bootstrap and provisioning
+and guidelines on how each is developed will make it easier to add support to ClusterAPI as well as to
+share implementations across bootstrap providers.
+
+
+**Compatibility and API guarantees**
+
+ClusterAPI has been stable for multiple years and is in widespread production use.
+Proposals and implementations of this working group must ensure to not excert operational
+risk on existing integrations.
+Any changes in the kubeadm bootstrap provider in particular must uphold guarantees and
+expectations set by the current implementation and must continue to support current use
+cases OR must provide means for current use cases to continue functioning for an appropriate
+migration period, as well as define a clear path for migration.
+
+## Scope
+
+1. Propose architectural improvements that abstract system configuration with the goal of
+   reducing duplicate functionality in current implementations, easing development of new configuration
+   features, and simplifying integration of new configuration systems for node bootstrapping.
```

Comment:
How about wording this a bit more explicitly, to avoid some of the overloaded terms:
```suggestion
1. Propose architectural improvements to the Machine provisioning process that separate Kubernetes bootstrapping tools (kubeadm, microk8s) from generating OS-distribution-specific initialization configuration (cloud-init, ignition) to avoid duplicate implementations, ease development of new configuration features and simplify implementing new providers for both kinds of tools.
```

Reply:
@schrej I like this except "initialization configuration" because we already call the same thing "provisioning" elsewhere and I think we should try to have one term for each concern since this area of the project is confusing as it is.

Reply:
The term is also used to describe the entire process of bringing up a new Machine, including creating the infrastructure, configuring the OS and setting it up as a K8s node. That's why I tried avoiding it here.

Reply:
It's not just (necessarily) OS / distribution specific initialisation - e.g. with Ignition one could perform significant customisation like setting the root partition filesystem, creating / adding partitions, using encrypted FS / encrypted volumes, and the like. Ignition runs in the initrd so supports significant customisation options.

I'd like to stick with "OS provisioning" for the time being, will update the section accordingly.